### PR TITLE
Bluetooth: Host: Add LE Power Control Request Procedure APIs

### PR DESCRIPTION
**Pull request by KyraLengfeld**
This commit adds the LE API's for the LE Power Control Request Feature in Zephyr.

The support of feature is provided with the controller-based feature selection with BT_CTLR_LE_POWER_CONTROL_SUPPORT and is selectable via BT_TRANSMIT_POWER_CONTROL.

With the new APIs, the applications will:
get improved reading of local and remote tx power
be aware of changes in remote or local tx power

Defined HCI commands in Core Spec v5.4:
7.8.117 LE Enhanced Read Transmit Power Level command: improvement to existing local tx power reading.
7.8.118 LE Read Remote Transmit Power Level command: Remote tx power is read through an event (LE Transmit Power Reporting) 7.8.121 LE Set Transmit Power Reporting Enable command: Enables local or remote tx power reporting to monitor changes in tx power 7.7.65.33 LE Transmit Power Reporting event

Note: to utilize the Feature fully, Nordic-LL-only vendor-specific commands are needed. These will not be added in RTOS zephyr but instead implemented in a maintainable way in sdk.

### CODE REVIEW

File: include/zephyr/bluetooth/conn.h
Comment (alwa-nordic) [inline include/zephyr/bluetooth/conn.h:1147]: Since there the same object is passed to multiple listeners, we should enforce constness.

```suggestion
				      const struct bt_conn_le_tx_power_report *report);
```

File: subsys/bluetooth/controller/Kconfig
Comment (alwa-nordic) [inline subsys/bluetooth/controller/Kconfig:493]: There is both a `default` here and a `select` in `BT_TRANSMIT_POWER_CONTROL`. Is it correct to have both?

**Review by alwa-nordic — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/Kconfig
Comment (KyraLengfeld) [reply subsys/bluetooth/controller/Kconfig:493]: You are right, best case overkill, worst case source of kconfig issues in future. I will remove the "default y if".

**Review by alwa-nordic — DISMISSED** [review]

File: subsys/bluetooth/host/Kconfig
Comment (wiba-nordic) [inline subsys/bluetooth/host/Kconfig:758]:
```suggestion
	  support LE Power Control Request feature that is defined in the Bluetooth
```

**Review by wiba-nordic — CHANGES_REQUESTED** [review]

**Review by wiba-nordic — APPROVED** [review]

**Review by hermabe — DISMISSED** [review]

**Review by alwa-nordic — DISMISSED** [review]

File: include/zephyr/bluetooth/conn.h
Comment (jhedberg) [inline include/zephyr/bluetooth/conn.h:1147]: minor: the split line doesn't seem to be aligned consistently (to the other similar places in this file, or Bluetooth subsystem code for that matter). Should be n tabs + 0-7 spaces to align right after the `(` of the previous line.

Comment (jhedberg) [reply include/zephyr/bluetooth/conn.h:1147]: FWIW, this seems to be the only place with this issue - the alignment looks correct in all other "split line" places in this PR

File: include/zephyr/bluetooth/conn.h
Comment (jhedberg) [inline include/zephyr/bluetooth/conn.h:489]: The public `conn.h` API isn't really tied to HCI in any way (although it might reuse defined values like I mentioned in my other comment), so I'm not sure a reference to a HCI command is appropriate.

Comment (KyraLengfeld) [reply include/zephyr/bluetooth/conn.h:489]: I can reference `bt_conn_le_get_remote_tx_power_level` instead? Otherwise I can describe it.

File: include/zephyr/bluetooth/conn.h
Comment (KyraLengfeld) [reply include/zephyr/bluetooth/conn.h:491]: I feel it is beneficial to have a description of the values defined like this, so I would just add a comment referencing the HCI specification? Thinking about your other comment this would be a bit contradictory to reference though(as conn.h is not tied to hci).

File: include/zephyr/bluetooth/conn.h
Comment (KyraLengfeld) [reply include/zephyr/bluetooth/conn.h:1147]: Will change soon. Thanks :)

File: include/zephyr/bluetooth/conn.h
Comment (jhedberg) [reply include/zephyr/bluetooth/conn.h:491]: > I feel it is beneficial to have a description of the values defined like this, so I would just add a comment referencing the HCI specification? Thinking about your other comment this would be a bit contradictory to reference though(as conn.h is not tied to hci).

I agree, it's beneficial. There should at least be a comment for the `reason` struct member that it contains HCI values. As for the values themselves, I'm thinking maybe `hci.h` is a better place for them (near the command response parameter definition) rather than `conn.h`?

File: include/zephyr/bluetooth/conn.h
Comment (jhedberg) [reply include/zephyr/bluetooth/conn.h:489]: > I can reference `bt_conn_le_get_remote_tx_power_level` instead? Otherwise I can describe it.

Either way is fine, thanks.

Comment (KyraLengfeld) [reply include/zephyr/bluetooth/conn.h]: I left that comment as is, as I shifted it into hci_types.h (Which I assume you meant in your other comment.).

File: include/zephyr/bluetooth/conn.h
Comment (KyraLengfeld) [reply include/zephyr/bluetooth/conn.h]: I added a core spec reference, imo core reference is the cleanest way to show where it comes from and the user can easily determine if it is outdated or not. If you though about something else, let me know :)

**Review by alwa-nordic — DISMISSED** [review]

**Review by hermabe — DISMISSED** [review]

File: include/zephyr/bluetooth/hci_types.h
Comment (jhedberg) [inline include/zephyr/bluetooth/hci_types.h:607]: I guess this should be moved closer to `struct bt_hci_evt_le_transmit_power_report` since that's where you have `uint8_t reason`? Also, since it's `uint8_t reason` and not `enum bt_conn_le_tx_power_report_reason` I'd stick to defines, like everything else in this header file, and the prefix would then be `BT_HCI_*` instead of `BT_CONN_*` or `bt_conn_*` (if you check, there aren't any other definitions in this header file that don't use the bt_hci prefix, so already from that perspective it's ill-suited here)

Comment (KyraLengfeld) [reply include/zephyr/bluetooth/hci_types.h:607]: I see your points. Reason why I placed it there was as we technically only use it in conn.h and in connection with read remote tx power. 
How about changing the report type to `enum bt_conn_le_tx_power_report_reason` (I would rename it to `enum bt_hci_le_tx_power_report_reason` ), actually using that in `bt_hci_evt_le_transmit_power_report` (as then it is defined right above, would be kind-off ugly imo to not utilize it), and keep utilizing it in `bt_conn_le_tx_power_report` in conn.h? Yes we would have an hci type (`enum bt_hci_le_tx_power_report_reason`) in conn.h, but we reference the hci spec regarding reporting reason above that parameter anyways.

Comment (KyraLengfeld) [reply include/zephyr/bluetooth/hci_types.h]: See the new changes for how I meant this.

File: include/zephyr/bluetooth/hci_types.h
Comment (jhedberg) [inline include/zephyr/bluetooth/hci_types.h:2953]: I don't think this is safe, and may not always work. Most platforms will use the native word width (e.g. 32 bits) for enum types. The only way to force the compiler to shrink the type to the minimum possible width is with the `__packed` annotation. Since these HCI structs themselves have a `__packed` annotation it *might* get "inherited" to the individual members, but I suspect it might also just be used to strip off unnecessary padding. Either way, this is not ideal: for these protocol structs we really do want to be explicit about the width of each member, so that the exactly match what's defined in the HCI specification, i.e. `uint8_t`, `uint16_t`, etc are really the appropriate types IMO.

Comment (KyraLengfeld) [reply include/zephyr/bluetooth/hci_types.h:2953]: Hm, yeah, I see. But then what do you think I should do about the enum? have it be a bt_conn enum, with BT_HCI types, that is defined in hci_types.h but only used in conn.h?

Comment (jhedberg) [reply include/zephyr/bluetooth/hci_types.h:2953]: I agree that enums are nice when they can be used, however in this case, if you want to be consistent between the HCI struct and the API, both would need to use `uint8_t`, and in that case you're back to my proposal of using `#define` :)

Comment (KyraLengfeld) [reply include/zephyr/bluetooth/hci_types.h:2953]: Alright :) In that case there will be no usage of these defines implemented though, don't see the benefit to have them defined then. Instead I'd just reference the spec and perhaps expand the explanation in the comment before the parameter. 
Let me know if I am missing something in that thought process.

Comment (jhedberg) [reply include/zephyr/bluetooth/hci_types.h:2953]: > Alright :) In that case there will be no usage of these defines implemented though, don't see the benefit to have them defined then. Instead I'd just reference the spec and perhaps expand the explanation in the comment before the parameter. Let me know if I am missing something in that thought process.

That's fine, though I do think they might be helpful to applications, so they don't need to hard-code values or create their own defines. In fact, this whole new feature submission should probably have been accompanied with a sample app that shows the intended usage of the new API. I.e. it's not just that there's no usage of the defines as the PR now stands, there's no usage of any part of the API, which is not really good when you consider of CI coverage and sample application code.

Comment (KyraLengfeld) [reply include/zephyr/bluetooth/hci_types.h:2953]: I agree with that, but the feature can only be utilized minimally with the le commands, the real benefit of it comes with the nordic-only vendor-specific commands.

**Comment by jhedberg** [comment]
Another thing I realized: it would be good to split the controller changes into a separate commit from the host ones. And if you add a sample application + CI integration (which I think you should) that should also be in a separate (third) commit.

File: include/zephyr/bluetooth/hci_types.h
Comment (jhedberg) [reply include/zephyr/bluetooth/hci_types.h:2953]: > I agree with that, but the feature can only be utilized minimally with the le commands, the real benefit of it comes with the nordic-only vendor-specific commands.

If there's no value to non-Nordic platforms, then the feature doesn't belong upstream. If there is some value, sure, but you need to demonstrate that the code actually works. Is building the new code currently covered by any upstream test cases? In other words, when I see the CI checks passing has it actually tried to build the new code? We also have the bsim tests which could be used to implement runtime CI checks for the feature and have CI verify that the code doesn't just build but actually works.

**Comment by KyraLengfeld** [comment]
> Another thing I realized: it would be good to split the controller changes into a separate commit from the host ones. And if you add a sample application + CI integration (which I think you should) that should also be in a separate (third) commit.

This is the LE part of changes, which we want eventually in the fork sdk-zephyr, but since the API's are basically just calling hci commands as defined by spec, it is not something sdk-specific and would expand rtos-zephyr's capabilities.
(See [Draft PR of sdk-nrf](https://github.com/nrfconnect/sdk-nrf/pull/13136) changes, the description/commit message should give a good summary of it.)

I did manual testing with bt shell (Which I can provide to you, or add as a commit for le commands here. Don't think that is needed to merge to main though).
There is also the integration of the enhanced get tx power command [here](https://github.com/zephyrproject-rtos/zephyr/pull/65273/files#diff-00881f70fc7ec49b966fed79bd6a5a8ed5357599b241b5b029736bbc97683f83R2727), not sure of the tps service get's tested though.

A sample would be a way to showcase these LE cmds (reading local and reading remote tx power), yet the transmit power report can only be showcased by changing tx power, which I don't see any API for in zephyr (there some HCI VS commands, see [HCI_PWR_CTRL sample](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/bluetooth/hci_pwr_ctrl), which can be used with the LE Power Control Request Feature, but is not part of it.)
This is just my personal opinion, so feel free to correct me, but I don't see much worth implementing a sample that can only read tx power but doesn't act upon that info in any way.
An option to have a sample read local/remote tx power would be to include them in the HCI_PWR_CTRL sample, yet that is HCI and wouldn't use the bt_conn API's in this PR.

How to include CI integration in Zephyr is beyond my knowledge, so i'd need quite some pointers how-to. If there is a nice way to test this feature in integration, I am all for it though :)

File: include/zephyr/bluetooth/hci_types.h
Comment (jhedberg) [inline include/zephyr/bluetooth/hci_types.h:2947]: The HCI spec defines these in hexadecimal, so it looks a bit odd that you've decided to use decimal values. Just follow the same what the spec does:
Screenshot 2023-11-27 at 11 57 02

Comment (jhedberg) [reply include/zephyr/bluetooth/hci_types.h:2947]: And one small thing: add an empty line between the defines and the struct

**Comment by jhedberg** [comment]
@KyraLengfeld thanks for the updates! Mostly style & consistency related comments left - after that I'd be ready to approve. The inconsistency with using "tx" in some places and "transmit" in others bothers me a little bit as well, but not enough to block the PR (feel free to fix it though, if you want)

**Comment by KyraLengfeld** [comment]
> @KyraLengfeld thanks for the updates! Mostly style & consistency related comments left - after that I'd be ready to approve. The inconsistency with using "tx" in some places and "transmit" in others bothers me a little bit as well, but not enough to block the PR (feel free to fix it though, if you want)

Thanks for reviewing! I amended according to your comments :) Regarding TX vs Transmit, most (if not all) names of functions/variables/defines are from how the spec calls it, and it uses both TX and Transmit (example `TX_Power_Level_Flag` parameter and `HCI_LE_Transmit_Power_Reporting` event naming). If there is anything in particular that still bothers you, let me know.

**Review by jhedberg — DISMISSED** [review]

File: subsys/bluetooth/shell/bt.c
Comment (jhedberg) [inline subsys/bluetooth/shell/bt.c:131]: Sorry, I saw this already earlier, but forgot to comment on it: no camelCase in Zephyr, please - use underscore instead (not sure why you changed the style only for some specific functions?)

File: subsys/bluetooth/shell/bt.c
Comment (jhedberg) [inline subsys/bluetooth/shell/bt.c:148]: same here

File: subsys/bluetooth/shell/bt.c
Comment (jhedberg) [inline subsys/bluetooth/shell/bt.c:162]: and here

File: subsys/bluetooth/shell/bt.c
Comment (jhedberg) [inline subsys/bluetooth/shell/bt.c:189]: Add a code comment after the `#endif` so that it's easy to see which block this terminates without having to scroll up:
```c
#endif /* CONFIG_BT_TRANSMIT_POWER_CONTROL */
```

File: include/zephyr/bluetooth/hci_types.h
Comment (jhedberg) [inline include/zephyr/bluetooth/hci_types.h:2953]: This shouldn't be signed, so use `uint8_t`. The power level and delta can be negative, but not the flags.

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/shell/bt.c
Comment (KyraLengfeld) [reply subsys/bluetooth/shell/bt.c:131]: I did so because I based it on the phy2str function, which isn't exactly camelCase but something in my mind thought if it had more words, it would be. I'll clean up what you commented :)

**Comment by jhedberg** [comment]
@KyraLengfeld looks generally good now, thanks! However, it still doesn't look like there's any sample or test that would actually build the new code in upstream CI. I did a quick local compilation test by forcing the new Kconfig option to be enabled, and was able to verify that the code builds without any warnings or errors. However, it would be important to have something in upstream CI doing these checks as well, so that we can be sure that the code keeps building cleanly as new changes go into Zephyr's Bluetooth stack.

**Comment by KyraLengfeld** [comment]
> @KyraLengfeld looks generally good now, thanks! However, it still doesn't look like there's any sample or test that would actually build the new code in upstream CI. I did a quick local compilation test by forcing the new Kconfig option to be enabled, and was able to verify that the code builds without any warnings or errors. However, it would be important to have something in upstream CI doing these checks as well, so that we can be sure that the code keeps building cleanly as new changes go into Zephyr's Bluetooth stack.

Thanks! 
Would such a test be integrated into https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/bluetooth/tester ?

And would there be the option to merge it in separate PR's? (As then while I figure out the tests in rtos zephyr, I can already get going on the sdk-PRs)

**Comment by jori-nordic** [comment]
@KyraLengfeld Just building the shell code that you added should be enough (right @jhedberg ?)

look at this: https://github.com/jori-nordic/zephyr/commit/d7c4ab605987025efb7dcd12bce8c612e435c59d

**Review by jhedberg — APPROVED** [review]
Assuming CI passes, this is finally good to be merged, at least from my side :)

**Review by jori-nordic — APPROVED** [review]

**Review by hermabe — APPROVED** [review]

**Pull request merged**
